lib/spack/spack/cmd/container.py

```diff
@@ -0,0 +1,151 @@
+             'svn', 'tar', 'tcl', 'unzip', 'zlib-devel'])
+
+    # GNU compilers
+    recipe += hpccm.building_blocks.gnu()
+
+    # Spack itself
+    spack_bb = bb_base()
+    spack_bb += comment('Spack')
+    spack_bb += shell(commands=[
+        'mkdir -p /opt && cd /opt',
+        'git clone --depth=1 --branch {} https://github.com/spack/spack'.format(args.spack_branch),
+        '/opt/spack/bin/spack bootstrap',
+        'ln -s /opt/spack/share/spack/setup-env.sh /etc/profile.d/spack.sh',
+        'ln -s /opt/spack/share/spack/spack-completion.bash /etc/profile.d'])
+    spack_bb += environment(variables={'PATH': '/opt/spack/bin:$PATH',
+                                       'FORCE_UNSAFE_CONFIGURE': '1'})
+    recipe += spack_bb
+
+    # spack install command line arguments
+    install_args = ['-y']
+    if args.no_checksum:
+        install_args.append('--no-checksum')
+    if args.show_log_on_error:
+        install_args.append('--show-log-on-error')
+    install_args_str = ' '.join(install_args)
+
+    # Process each spec
+    for spec in spack.cmd.parse_specs(args.specs):
+        spec.concretize()
+
+        bb = bb_base()
+        bb += comment('{0} {1}'.format(spec.name, spec.version))
+
+        # Process each dependency.  Use a selectable depth to balance
+        # the build cache vs. container spec file complexity
+        for depth, dspec in spec.traverse_edges(order='post', depth=True):
+            if depth <= args.depth:
+                node = dspec.spec
+                cmd = 'spack install {0} {1}@{2}'.format(install_args_str,
+                                                         node.name,
+                                                         node.version)
+                bb += shell(chdir=False, commands=[cmd, 'spack clean --all'])
+
+        recipe += bb
+
+    hpccm.config.set_container_format(args.format)
+    print(recipe)
+
+    # Really basic implementation of a multi-stage container
+    # specification.  Assume the same base image should be used and
+    # just copy /opt/spack.  If more advanced runtime capabilities
+    # become available, this should become more selective in what
+    # files are copied for a much bigger reduction in container size
+    # (e.g., build vs. link dependencies).  Multi-stage containers are
+    # Docker specific.
+    if args.multi_stage and args.format == 'docker':
+        runtime = hpccm.Stage()
+        runtime += baseimage(image=args.baseimage, _distro=args.distro)
+        runtime += recipe.runtime()
+
+        spack_rt = bb_base()
+        spack_rt += comment('Spack')
+        spack_rt += copy(_from='0', src='/opt/spack', dest='/opt/spack')
+        spack_rt += shell(commands=[
+            'ln -s /opt/spack/share/spack/setup-env.sh'
+            ' /etc/profile.d/spack.sh',
+            'ln -s /opt/spack/share/spack/spack-completion.bash'
+            ' /etc/profile.d'])
+        spack_rt += environment(variables={'PATH': '/opt/spack/bin:$PATH'})
```

Comment:
This will be redundant in some cases, but there are also container cases where the shell start files are not invoked.  

For instance, SPACK_ROOT is not set when directly invoking commands, but is when starting a shell in the container:
```
$ docker run --rm -it spack ls -l /etc/profile.d/spack.sh
lrwxrwxrwx. 1 root root 35 Apr 22 16:37 /etc/profile.d/spack.sh -> /opt/spack/share/spack/setup-env.sh
$ docker run --rm -it spack printenv SPACK_ROOT
$ docker run --rm -it spack printenv PATH
/opt/spack/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin

$ docker run --rm -it spack
[root@c3bdf833b061 /]# printenv SPACK_ROOT
/
```